trl/trainer/sft_trainer.py

```diff
@@ -0,0 +1,244 @@
+        args: TrainingArguments = None,
+        data_collator: Optional[DataCollator] = None,
+        train_dataset: Optional[Dataset] = None,
+        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        model_init: Optional[Callable[[], PreTrainedModel]] = None,
+        compute_metrics: Optional[Callable[[EvalPrediction], Dict]] = None,
+        callbacks: Optional[List[TrainerCallback]] = None,
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+        peft_config: Optional[Dict] = None,
+        dataset_text_field: Optional[str] = None,
+        packing: Optional[bool] = True,
+        dataset_kwargs: Optional[Dict] = {},
+        prepare_in_int8_kwargs: Optional[Dict] = {},
+        **pretrained_kwargs,
+    ):
+        if isinstance(model, str):
+            warnings.warn(
+                "You passed a model_id and a `PeftConfig` to the SFTTrainer. This will automatically create an "
+                "`AutoModelForCausalLM` or a `PeftModel` for you."
+            )
+
+        if is_peft_available() and peft_config is not None:
+            if not isinstance(peft_config, PeftConfig):
+                raise ValueError(
+                    "If you want to use the PeftModel, you need to pass a PeftConfig object to the SFTTrainer."
+                    f" and you passed a {type(peft_config)}."
+                )
+
+            if not isinstance(model, PeftModel):
+                if pretrained_kwargs.get("load_in_8bit", False) and "device_map" not in pretrained_kwargs:
+                    device_map = {"": Accelerator().process_index}
+                    pretrained_kwargs["device_map"] = device_map
+
+                model = AutoModelForCausalLM.from_pretrained(
+                    model,
+                    **pretrained_kwargs,
+                )
+
+                if getattr(model, "is_loaded_in_8bit", False):
+                    model = prepare_model_for_int8_training(model, **prepare_in_int8_kwargs)
+
+                model = get_peft_model(model, peft_config)
+
+            if callbacks is None:
+                callbacks = [PeftSavingCallback]
+
+        elif not isinstance(model, PreTrainedModel):
+            model = AutoModelForCausalLM.from_pretrained(model, **pretrained_kwargs)
+
+        if tokenizer is None:
+            tokenizer = AutoTokenizer.from_pretrained(model.config._name_or_path)
+            if getattr(tokenizer, "pad_token", None) is None:
+                tokenizer.pad_token = tokenizer.eos_token
+
+        # check if torch dataset / dataloader and do nothing
+        if train_dataset is not None and (
+            isinstance(
+                train_dataset,
+                (torch.utils.data.IterableDataset, torch.utils.data.Dataset),
+            )
+        ):
```

Comment:
don't we need the same test for eval?

Reply:
I changed the way we create the datasets a bit!